src/EditorFeatures/Text/Extensions.SnapshotSourceText.cs

```diff
@@ -262,6 +239,53 @@ public override SourceText WithChanges(IEnumerable<TextChange> changes)
                     currentSnapshot: ((ITextSnapshot2)buffer.CurrentSnapshot).TextImage);
             }
 
+            private static ITextImage RecordReverseMapAndGetImage(ITextSnapshot editorSnapshot)
+            {
+                Contract.ThrowIfNull(editorSnapshot);
+
+                var textImage = ((ITextSnapshot2)editorSnapshot).TextImage;
+                Contract.ThrowIfNull(textImage);
+
+                // If we're already in the map, there's nothing to update.  Do a quick check
+                // to avoid two allocations per call to RecordTextSnapshotAndGetImage.
+                if (!s_roslynToEditorSnapshotMap.TryGetValue(textImage, out var weakReference))
+                {
+                    // put reverse entry that won't hold onto anything
+                    weakReference = s_roslynToEditorSnapshotMap.GetValue(
+                        textImage, _ => new WeakReference<ITextSnapshot>(editorSnapshot));
+                }
+
+#if DEBUG
+                // forward and reversed map is 1:1 map. as long as snapshot is not null, it can't be
+                // different
+                var snapshot = weakReference.GetTarget();
+                Contract.ThrowIfFalse(snapshot == null || snapshot == editorSnapshot);
```

Comment:
The `snapshot == null` part seems bad here, as I suspect had this been enabled we would have caught this bug long ago: if that is ever true, it means s_roslynToEditorSnapshotMap should have been updated and we've retriggered this bug. Should we remove this condition?

Reply:
snapshot can be null. if someone did var sourceText = ITextSnapshot(#1).AsSourceText(); and let ITextSnapshot(#1) go. then, sourceText.TryGetCorresponding... will return null (snapshot == null)

Reply:
Yes, but this function isn't called there. :-) The weak reference can obviously be lost (otherwise what's the point), but it should never disappear in this function.

Imagine we took the case where the if branch was taken, and we did create a weak reference for the text image that was passed. In debug, we have GC roots for editorSnapshot (because of this line) and textImage (because it's going to be returned), therefore, the weak reference can't possibly have been collected yet. Thus, the latter assert is acceptable.

Now consider the case where the branch wasn't taken; that means that the textImage -> textSnapshot weak entry has already been created. By your statement to my earlier question, there is a 1:1 mapping here, so if the textImage is alive, the it must be the same snapshot as editorSnapshot, which is also alive (because it was handed to this function). Therefore, it can't be null. If it *was* null, that means there are two different snapshots with the same TextImage, which invalidates your assertion when you linked to BaseSnapshot.cs earlier, and if so, this entire approach is broken.

Follow me?

Reply:
ah. yes. it can't be null :)